README.md
# RIF Storage.js Pinning service

[![CircleCI](https://flat.badgen.net/circleci/github/rsksmart/rif-storage-pinner/master)](https://circleci.com/gh/rsksmart/rif-storage-pinner/)
[![Dependency Status](https://david-dm.org/rsksmart/rif-storage-pinner.svg?style=flat-square)](https://david-dm.org/rsksmart/rif-storage-pinner)
[![](https://img.shields.io/badge/made%20by-IOVLabs-blue.svg?style=flat-square)](http://iovlabs.org)
[![](https://img.shields.io/badge/project-RIF%20Storage-blue.svg?style=flat-square)](https://www.rifos.org/)
[![standard-readme compliant](https://img.shields.io/badge/standard--readme-OK-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
[![Managed by tAsEgir](https://img.shields.io/badge/%20managed%20by-tasegir-brightgreen?style=flat-square)](https://github.com/auhau/tasegir)
![](https://img.shields.io/badge/npm-%3E%3D6.0.0-orange.svg?style=flat-square)
![](https://img.shields.io/badge/Node.js-%3E%3D10.0.0-orange.svg?style=flat-square)

> Application for providing your storage space on decentralized storage networks to other to use in exchange of RIF Tokens

## Table of Contents

- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [License](#license)

## Usage

### Running on bare metal

Install the package from NPM:
```bash
$ npm install @rsksmart/rif-storage-pinning
```

When you initialize the Pinning service you need Offer ID. It is address of your account that you will use to create
the Offer on the RIF Marketplace. You can either go to the RIF Marketplace directly and follow its guidance on how to set up
new Offer, or if you need to set up the Pinning service prior that just use the account address instead here.

Example of usage:
```bash
$ rif-pinning init --offerId=0x123456789
$ rif-pinning daemon --provider='ws://localhost:8546' --ipfs='http://localhost:5001' --network=testnet --strategy=blockchain
```

This will:
 - initialize pinning service for given Offer
 - start the daemon of the pinning service
 - listens only for events for the Offer ID `0x123456789`
 - use blockchain node for listening on events at `ws://localhost:8546` that is connected to testnet network
 - thanks to `--network testnet` will use predefined deployed smart-contracts on testnet
 - connects to your locally running IPFS node at `http://localhost:5001`

### Docker

Currently, the Docker image is not published anywhere, so you have to clone/download this repository first.
Bellow we will describe only deployment using `docker-compose`, but it is possible to use only Docker image as well.

In the root of the repository create file `.env-pinner` where you can specify any environmental variable as specified bellow. Example is:

```
RIFS_OFFER='' # fill in the Offer ID here
RIFS_STRATEGY=blockchain
RIFS_NETWORK=testnet
RIFS_PROVIDER='' # fill in the address of RSKj node
RIFS_COMMS_BOOTSTRAP='' # fill in the address of libp2p bootstrap nodes
```

Then to start the pinning service you can simply run:

```bash
$ docker-compose up
```

## Environmental variables

Pinning service supports following environmental variables:

 - `RIFS_CONFIG` (`string`) - Same like `--config` flag.
 - `RIFS_MARKETPLACE_URL` (`string`) - Marketplace URL
 - `RIFS_CONTRACT_ADDR` (`string`) - Specifies address of smart contract to listen the events from. Mainly for development as this is otherwise configured using `--network` flag.
 - `RIFS_COMMS_TRANSPORT` (`cache`/`libp2p`) - Defines the transport for comms
 - `RIFS_COMMS_BOOTSTRAP_ENABLED` (`true`/`false`) - Defines if bootstrap should be used. Same as libp2p config's [`bootstrap.enabled`](https://github.com/libp2p/js-libp2p-bootstrap) property.
 - `RIFS_COMMS_BOOTSTRAP_LIST` (`array`) - Defines an array of multiaddress that the Pinner's libp2p node will use to bootstrap its connectivity. Same as libp2p config's [`bootstrap.list`](https://github.com/libp2p/js-libp2p-bootstrap) property.
 - `RIFS_DB` (`string`) - Specify the name or path to the data base file.
 - `RIFS_IPFS` (`string`) - Same like `--ipfs` flag.
 - `RIFS_NETWORK` (`testnet|mainnet`) - Same like `--network` flag.
 - `RIFS_OFFER` (`string`) - Specifies Offer Id which the Pinning service should listen on. Same like `--offerId` flag.
 - `RIFS_STRATEGY` (`blockchain|marketplace`) - Same like `--strategy` flag.
 - `RIFS_PROVIDER` (`string`) - Same like `--provider` flag.
 - `LOG_LEVEL` (`string`) - Same like `--log` flag.
 - `LOG_FILTER` (`string`) - Same like `--log-filter` flag.
 - `LOG_PATH` (`string`) - Same like `--log-path` flag.
 - `LOG_NO_COLORS` (boolean) - if set the output won't be colorized

## Commands
<!-- commands -->
* [`rif-pinning agreements`](#rif-pinning-agreements)
* [`rif-pinning cleanup`](#rif-pinning-cleanup)
* [`rif-pinning daemon`](#rif-pinning-daemon)
* [`rif-pinning db-migration`](#rif-pinning-db-migration)
* [`rif-pinning init`](#rif-pinning-init)

### `rif-pinning agreements`

Agreements info

```
USAGE
  $ rif-pinning agreements

OPTIONS
  -d, --db=db                                                   Name or path to DB file
  -p, --pinningStatus=running|backoff|created|finished|errored  Filter by pinning status
  -s, --status=active|inactive                                  Filter by status
  --config=config                                               path to JSON config file to load
  --log=error|warn|info|verbose|debug                           [default: error] what level of information to log
  --log-filter=log-filter                                       what components should be logged (+-, chars allowed)
  --log-path=log-path                                           log to file, default is STDOUT
  --skipPrompt                                                  Answer yes for any prompting

EXAMPLES
  $ rif-pinning agreements
  $ rif-pinning agreements --db myOffer.sqlite
  $ rif-pinning agreements --ls -f active
  $ rif-pinning agreements --ls -f inactive
  $ rif-pinning agreements --ls -f inactive -p pinned
  $ rif-pinning agreements --ls -f active -p not-pinned
```

### `rif-pinning cleanup`

Cleanup pinner files

```
USAGE
  $ rif-pinning cleanup

OPTIONS
  -d, --db=db                          Name or path to DB file
  -u, --unpin                          Unpin all files
  --config=config                      path to JSON config file to load
  --log=error|warn|info|verbose|debug  [default: error] what level of information to log
  --log-filter=log-filter              what components should be logged (+-, chars allowed)
  --log-path=log-path                  log to file, default is STDOUT
  --skipPrompt                         Answer yes for any prompting

EXAMPLES
  $ rif-pinning cleanup
  $ rif-pinning cleanup --db myOffer.sqlite
  $ rif-pinning cleanup --unpin
```

### `rif-pinning daemon`

Run pinning service

```
USAGE
  $ rif-pinning daemon

OPTIONS
  -d, --db=db                          Name or path to DB file
  -n, --network=testnet|mainnet        specifies to which network is the provider connected
  -p, --provider=provider              URL to blockchain node or Marketplace server
  --config=config                      path to JSON config file to load

  --ipfs=ipfs                          specifies a connection URL to IPFS node. Default is go-ipfs listening
                                       configuration.

  --log=error|warn|info|verbose|debug  [default: error] what level of information to log

  --log-filter=log-filter              what components should be logged (+-, chars allowed)

  --log-path=log-path                  log to file, default is STDOUT

  --skipPrompt                         Answer yes for any prompting

  --strategy=marketplace|blockchain    what type of provider will be used for listening on events. Default is
                                       "marketplace". For blockchain you have to have access to a node that has allowed
                                       eth_getLogs call.

EXAMPLES
  $ rif-pinning daemon --strategy=blockchain --provider 'ws://localhost:8546' --ipfs '/ip4/127.0.0.1/tcp/5001' --network
  testnet

  $ rif-pinning daemon --strategy=marketplace --ipfs 'http://localhost:5001' --network testnet
```

### `rif-pinning db-migration`

DB migration

```
USAGE
  $ rif-pinning db-migration

OPTIONS
  -d, --db=db                          Name or path to DB file
  -d, --down                           Undo db migration
  -d, --generate=generate              Generate migrations using template [--generate=migration_name]
  -m, --migration=migration            Migration file
  -t, --to=to                          Migrate to
  -u, --up                             Migrate DB
  --config=config                      path to JSON config file to load
  --log=error|warn|info|verbose|debug  [default: error] what level of information to log
  --log-filter=log-filter              what components should be logged (+-, chars allowed)
  --log-path=log-path                  log to file, default is STDOUT
  --skipPrompt                         Answer yes for any prompting

EXAMPLES
  $ rif-pinning db --up
  $ rif-pinning db --down
  $ rif-pinning db --up --to 0-test
  $ rif-pinning db --up --migration 01-test --migration 02-test
  $ rif-pinning db --up --db ./test.sqlite --to 09-test
  $ rif-pinning db --down --db ./test.sqlite --to 09-test
  $ rif-pinning db --generate my_first_migration
```

### `rif-pinning init`

Initialize Pinner service dependencies

```
USAGE
  $ rif-pinning init

OPTIONS
  -d, --db=db                          Name or path to DB file
  -o, --offerId=offerId                ID of Offer to which should the service listen to
  -s, --keySize=keySize                [default: 2048] Size of private key that will be used for Peer Identity
  -t, --keyType=rsa|ed25519|secp256k1  [default: rsa] Type of private key that will be used for Peer Identity
  --config=config                      path to JSON config file to load
  --log=error|warn|info|verbose|debug  [default: error] what level of information to log
  --log-filter=log-filter              what components should be logged (+-, chars allowed)
  --log-path=log-path                  log to file, default is STDOUT
  --[no-]override-db                   Skip the prompt when database exists with used value --override-db/--no-override-db
  --skipPrompt                         Answer yes for any prompting

EXAMPLES
  $ rif-pinning init
  $ rif-pinning init --offerId 0x123 --db ./relativeOrAbsolutePath/db.sqlite
  $ rif-pinning init --db fileName.sqlite
  $ rif-pinning init --db ./folder
```
<!-- commandsstop -->

## Contribute

There are some ways you can make this module better:

- Consult our [open issues](https://github.com/rsksmart/rif-storage-pinner/issues) and take on one of them
- Help our tests reach 100% coverage!

### Development

**Requirements:**

 * IPFSv5 and higher
 * Ganache
 * Node & NPM

Please on how to set up the development environment see [Development guide](./DEVELOPMENT.md)

#### Tips

 - Using the `npm run ipfs:consumer` and `npm run ipfs:provider` you can interact with
 each IPFS node using the standard commands that IPFS supports.
 - You can interact with the CLI using `npm run bin` script from the local folder.
 - Use [RIF Communication PubSub Node](https://github.com/rsksmart/rif-communications-pubsub-node) to listen on
 the broadcast events! Only configure the correct Room's name!

## License

[MIT](./LICENSE)


CHANGELOG.md
<a name="0.2.4"></a>
## [0.2.4](https://github.com/rsksmart/rds-ipfs/compare/v0.2.3-next.1...v0.2.4) (2021-04-22)


### Features

* **devops:** fixes on cd workflow for security ([#406](https://github.com/rsksmart/rds-ipfs/issues/406)) ([3dfdc95](https://github.com/rsksmart/rds-ipfs/commit/3dfdc95))
* **fix:** provider flag for daemon command ([#428](https://github.com/rsksmart/rif-storage-pinner/pull/428))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/rsksmart/rds-ipfs/compare/v0.2.3-next.1...v0.2.3) (2021-04-19)

* Fix `--network` flag for CLI
* adjust configs



<a name="0.2.2"></a>
## [0.2.2](https://github.com/rsksmart/rds-ipfs/compare/v0.2.1...v0.2.2) (2021-03-04)

* update `@rsksmart/rif-marketplace-storage` to `0.1.0`

<a name="0.2.1"></a>
## [0.2.1](https://github.com/rsksmart/rds-ipfs/compare/v0.2.0...v0.2.1) (2021-02-01)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/rsksmart/rds-ipfs/compare/v0.1.2...v0.2.0) (2021-01-21)


### Bug Fixes

* daemon command provider flag ([#338](https://github.com/rsksmart/rds-ipfs/issues/338)) ([b22730e](https://github.com/rsksmart/rds-ipfs/commit/b22730e))
* release command ([6b2b1e9](https://github.com/rsksmart/rds-ipfs/commit/6b2b1e9))


### Features

* comms marketplace transport ([#323](https://github.com/rsksmart/rds-ipfs/issues/323)) ([22d4136](https://github.com/rsksmart/rds-ipfs/commit/22d4136))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/rsksmart/rds-ipfs/compare/v0.1.1...v0.1.2) (2020-12-21)


### Bug Fixes

* init command does not fail when rolling back ([5c0ef1b](https://github.com/rsksmart/rds-ipfs/commit/5c0ef1b))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.6...v0.1.1) (2020-12-18)


### Bug Fixes

* send agreement new message before start pinning ([#276](https://github.com/rsksmart/rds-ipfs/issues/276)) ([0bb1b21](https://github.com/rsksmart/rds-ipfs/commit/0bb1b21))


### Features

* add event transformer ([#290](https://github.com/rsksmart/rds-ipfs/issues/290)) ([061119e](https://github.com/rsksmart/rds-ipfs/commit/061119e))
* lower case offer-id in topic computation ([#285](https://github.com/rsksmart/rds-ipfs/issues/285)) ([1c299e6](https://github.com/rsksmart/rds-ipfs/commit/1c299e6))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.6...v0.1.0) (2020-12-14)


### Bug Fixes

* send agreement new message before start pinning ([#276](https://github.com/rsksmart/rds-ipfs/issues/276)) ([0bb1b21](https://github.com/rsksmart/rds-ipfs/commit/0bb1b21))


### Features

* add event transformer ([#290](https://github.com/rsksmart/rds-ipfs/issues/290)) ([061119e](https://github.com/rsksmart/rds-ipfs/commit/061119e))
* lower case offer-id in topic computation ([#285](https://github.com/rsksmart/rds-ipfs/issues/285)) ([1c299e6](https://github.com/rsksmart/rds-ipfs/commit/1c299e6))



<a name="0.1.0-pre.6"></a>
# [0.1.0-pre.6](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.4...v0.1.0-pre.6) (2020-11-27)


### Bug Fixes

* **core:** logs colors handling ([#265](https://github.com/rsksmart/rds-ipfs/issues/265)) ([4bca600](https://github.com/rsksmart/rds-ipfs/commit/4bca600))
* fix bytes to megabytes convertion ([#257](https://github.com/rsksmart/rds-ipfs/issues/257)) ([e8fef55](https://github.com/rsksmart/rds-ipfs/commit/e8fef55))
* make contract address lower case for room topic ([#258](https://github.com/rsksmart/rds-ipfs/issues/258)) ([0db3317](https://github.com/rsksmart/rds-ipfs/commit/0db3317))
* room topic calculation ([#242](https://github.com/rsksmart/rds-ipfs/issues/242)) ([459f756](https://github.com/rsksmart/rds-ipfs/commit/459f756))


### Features

* check actual size after pin ([#264](https://github.com/rsksmart/rds-ipfs/issues/264)) ([4c148d3](https://github.com/rsksmart/rds-ipfs/commit/4c148d3))
* report when app running ([#267](https://github.com/rsksmart/rds-ipfs/issues/267)) ([a5fcebc](https://github.com/rsksmart/rds-ipfs/commit/a5fcebc))



<a name="0.1.0-pre.5"></a>
# [0.1.0-pre.5](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.4...v0.1.0-pre.5) (2020-11-18)


### Bug Fixes

* room topic calculation ([#242](https://github.com/rsksmart/rds-ipfs/issues/242)) ([459f756](https://github.com/rsksmart/rds-ipfs/commit/459f756))



<a name="0.1.0-pre.4"></a>
# [0.1.0-pre.4](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.3...v0.1.0-pre.4) (2020-11-16)


### Bug Fixes

* basic timeout estimation ([d231ca2](https://github.com/rsksmart/rds-ipfs/commit/d231ca2))
* privKey possible undefined ([6db9a7e](https://github.com/rsksmart/rds-ipfs/commit/6db9a7e))
* privKey possible undefined in tests ([#215](https://github.com/rsksmart/rds-ipfs/issues/215)) ([e464ae1](https://github.com/rsksmart/rds-ipfs/commit/e464ae1))


### Features

* **comms:** rebroadcast past messages ([#193](https://github.com/rsksmart/rds-ipfs/issues/193)) ([b6ac0e0](https://github.com/rsksmart/rds-ipfs/commit/b6ac0e0))



<a name="0.1.0-pre.3"></a>
# [0.1.0-pre.3](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.2...v0.1.0-pre.3) (2020-08-28)


### Features

* communication integration ([#159](https://github.com/rsksmart/rds-ipfs/issues/159)) ([8635d09](https://github.com/rsksmart/rds-ipfs/commit/8635d09))
* direct swarming ([#185](https://github.com/rsksmart/rds-ipfs/issues/185)) ([5590697](https://github.com/rsksmart/rds-ipfs/commit/5590697))
* marketplace strategy reorgs handling ([#162](https://github.com/rsksmart/rds-ipfs/issues/162)) ([9002fc4](https://github.com/rsksmart/rds-ipfs/commit/9002fc4))



<a name="0.1.0-pre.2"></a>
# [0.1.0-pre.2](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.1...v0.1.0-pre.2) (2020-08-21)


### Bug Fixes

* migrate also js migrations ([553560b](https://github.com/rsksmart/rds-ipfs/commit/553560b))



<a name="0.1.0-pre.1"></a>
# [0.1.0-pre.1](https://github.com/rsksmart/rds-ipfs/compare/v0.1.0-pre.0...v0.1.0-pre.1) (2020-08-21)



<a name="0.1.0-pre.0"></a>

src/cli.ts
import { Command, flags } from '@oclif/command'
import type { Input, OutputFlags } from '@oclif/parser'
import config from 'config'
import path from 'path'

import initApp from './'
import { Config, Strategy } from './definitions'

export default class PinningServiceCommand extends Command {
  static description = `
Pinning Service that is part of RIF Storage.

This service is needed to provide your storage space as part of RIF Marketplace. It listens on events and when there is new Agreement for specified Offer it will pin the content to your configured IPFS node.

By default it uses RIF Marketplace servers to listen on events, which are based on events from blockchain. You can eliminate this middle-man component and listen to events directly from blockchain. For that use --strategy=blockchain, but you have to also provide an blockchain node that has enabled eth_getLogs call using the --provider flag.
`
  static examples = [
    '$ rif-pinning --offerId 0x123456789 --strategy=blockchain --provider \'ws://localhost:8546\' --ipfs \'/ip4/127.0.0.1/tcp/5001\' --network testnet',
    '',
    '$ rif-pinning --offerId 0x123456789 --strategy=marketplace --ipfs \'/ip4/127.0.0.1/tcp/5001\' --network testnet'
  ]

  static flags = {
    offerId: flags.string({
      char: 'o',
      description: 'ID of Offer to which should the service listen to',
      env: 'RIFS_OFFER',
      required: true
    }),
    network: flags.string({
      char: 'n',
      description: 'specifies to which network is the provider connected',
      options: ['testnet', 'mainnet'],
      env: 'RIFS_NETWORK'
    }),
    provider: flags.string({
      char: 'p',
      description: 'URL to blockchain node or Marketplace server',
      env: 'RIFS_PROVIDER'
    }),
    strategy: flags.string({
      description: 'what type of provider will be used for listening on events. Default is "marketplace". For blockchain you have to have access to a node that has allowed eth_getLogs call.',
      options: ['marketplace', 'blockchain']
    }),
    'remove-cache': flags.boolean({
      description: 'removes the local database prior running the service'
    }),
    config: flags.string({
      description: 'path to JSON config file to load',
      hidden: true,
      env: 'RIFS_CONFIG'
    }),
    ipfs: flags.string({
      description: 'specifies a connection URL to IPFS node. Default is go-ipfs listening configuration.',
      env: 'RIFS_IPFS'
    }),
    log: flags.string({
      description: 'what level of information to log',
      options: ['error', 'warn', 'info', 'verbose', 'debug'],
      default: 'error',
      env: 'LOG_LEVEL'
    }),
    'log-filter': flags.string(
      {
        description: 'what components should be logged (+-, chars allowed)'
      }
    ),
    'log-path': flags.string(
      {
        description: 'log to file, default is STDOUT'
      }
    )
  }

  private configSetup (flags: OutputFlags<typeof PinningServiceCommand.flags>): void {
    const configObject: Config = {
      log: {
        level: flags.log,
        filter: flags['log-filter'] || null,
        path: flags['log-path'] || null
      }
    }

    let userConfig: Config = {}

    if (flags.config) {
      userConfig = config.util.parseFile(flags.config)
    }

    if (flags.strategy) {
      configObject.strategy = flags.strategy
    }

    if (flags.provider) {
      // We have to use hardcoded 'Strategy.Marketplace' here as default value, because we cant touch the `config` object yet.
      const strategy = userConfig.strategy ?? configObject.strategy ?? Strategy.Marketplace

      if (strategy === Strategy.Blockchain) {
        configObject.blockchain = { provider: flags.provider }
      } else {
        configObject.marketplace = { provider: flags.provider }
      }
    }

    if (flags.ipfs) {
      configObject.ipfs = { clientOptions: { url: flags.ipfs } }
    }

    config.util.extendDeep(config, userConfig)
    config.util.extendDeep(config, configObject)

    if (flags.network) {
      const networkConfigPath = path.join(__dirname, '..', 'config', `${flags.network}.json5`)
      config.util.extendDeep(config, config.util.parseFile(networkConfigPath))
    }

    if (!config.has('blockchain.contractAddress')) {
      throw new Error('You have to specify address of smart contract! Use --network flag!')
    }
  }

  async run () {
    const { flags: originalFlags } = this.parse(this.constructor as Input<typeof PinningServiceCommand.flags>)
    const flags = originalFlags as OutputFlags<typeof PinningServiceCommand.flags>
    this.configSetup(flags)

    await initApp(flags.offerId, { removeCache: Boolean(flags['remove-cache']), dataDir: this.config.dataDir })
  }
}
